Feature/add sorter param

callahat opened the pull request:
Adds support for passing  a sorter to `encode`
Specs ran against elixir 1.14.0, erlang 23.3.4.20
and elixir elixir 1.18.3-otp-27, erlang 27.3.3

Sorter support for `List.keysort` was added in elixir 1.14 - https://hexdocs.pm/elixir/1.14.0/List.html#keysort/3 hence the major version bump (in addition to an additional parameter being added to `StableJason.encode` functions)

callahat committed:
Adding support to pass in a sorter

callahat committed:
Additional spec for using as atom as the sorter

callahat committed:
Add sorter param to StableJason.encode

callahat committed:
Update README

callahat committed:
Slip the sorter into the opts param

egze commented on README.md:
```diff
@@ -14,13 +14,13 @@ It works similar like [OJSON](https://hex.pm/packages/ojson) but can output a pr
 {:ok, ~S|{"a":1,"b":2,"c":3}|}
 
-StableJason.encode(%{c: 3, b: 2, a: 1}, pretty: true)
+StableJason.encode(%{c: 3, b: 2, a: 1}, :asc, pretty: true)
```
```suggestion
StableJason.encode(%{c: 3, b: 2, a: 1}, sorter: :asc, pretty: true)
```

egze commented on README.md:
```diff
@@ -14,13 +14,13 @@ It works similar like [OJSON](https://hex.pm/packages/ojson) but can output a pr
 "{\"a\":1,\"b\":2,\"c\":3}"
 
-StableJason.encode!(%{c: 3, b: 2, a: 1}, pretty: true)
+StableJason.encode!(%{c: 3, b: 2, a: 1}, :asc, pretty: true)
```
```suggestion
StableJason.encode!(%{c: 3, b: 2, a: 1}, sorter: :asc, pretty: true)
```

callahat committed:
Update README.md
Update ci.yml to remove elixir 1.13 from the matrix, add elixir 1.17 and elixir 1.18

callahat committed:
mix format

egze commented:
@callahat Thank you for the PR. I'll merge and release a new version in couple of hours.

Just for my curiosity:

* How are you using StableJason?
* What is the use case for a custom sorter?

callahat commented:
@egze You're welcome.

I am using StableJason to prettify JSON being written to a file and checked in to a repo. The JSON is export data that's used to define a game (assets, scripts, maps, etc) written for a custom engine. Being able to have the object keys in a set order should cut down on some of the noise of things shifting about (without actually changing) in the pretty file, which should make it easier to diff between versions of a game to see what is different.

For the custom sorter, I wanted to be able to sort keys based on length then normal order (similar to how numbers are sorted with fewer digits being "higher"). Some of the keys in the JSON export mentioned above are really base 84 numbers - but to represent them the digits are strings - so hopefully this will also cut down a bit on keys shifting around.
